### ISSUE TITLE
state parameter in reducer function

### ISSUE THREAD

**Issue by sagswe** [aGuyNamedJonas/slim-redux#10]
I have slim-redux 0.0.11 and slim-redux-react 0.0.10 versions. my initial state is {counter:0} and i dont have existing roo reducer, so it was null.  The counter container snippet is like this
```
export default slimReduxReact({
  component: Counter,
  subscriptions: { counter: 'state.counter'},
  changeTriggers:{
    increaseCounter: {
      actionType: 'INCREMENT',
      reducer: ({counter}) => {
        debugger;
        return {counter:counter+ 1};
      }
    },
    decreaseCounter: {
      actionType: 'DECREMENT',
      reducer: (state) => {
        return {counter:state.counter - 1};
      }
    }
  }
});
```
i want to\he reducers to recieve whatever the subscription slice is . In my case subscription slice is 'state.counter'. But i am getting entire state. basically i would want to return counter +1 inside the reducer, instead of return {counter: counter+1}.

possible?. isn't that  reducers supposed to recieve their slice and operate on thier slice of state?

**Comment by aGuyNamedJonas** [comment]
Hey @sagswe, thanks for reaching out! What you're looking for is essentially the type of behaviour we're used to from working with root reducers right? It's interesting that you would bring this up, because I've also thought about how to replicate this kind of reducer which only modifies a slice of the state in slim-redux.

I'm currently working to complete the 0.2 release of slim-redux / slim-redux-react which comes with a complete API reimplementation (plus some really cool extensions to it) and feels like it will cure a lot of headaches I had with the first release. So right now is the perfect time to discuss requests like the one you brought up.  

Here's an idea for an addition to `changeTrigger()` which could accomplish what I think you're looking for:

```
export const addTodo = changeTrigger('ADD_TODO', 'state.todos', (label, todos) => {
  const newId = todos.map((max, todo) => Math.max(max, todo.id), 0) + 1;
  return [
      ...todos,
      {id: newId, label: label, checked: false},
    ]
});
```

So in this case `changeTrigger()` has an optional second subscription-style string which determines which part of the state tree the reducer function is modifying. In that case the part of the state that the change trigger is focused on will be passed in to the reducer function as the last argument (in this case "todos").  

Does this solution make sense? Or is there anything that you don't like about this solution / anything where you can forsee issues with this solution or would like a detail about it changed?  

Please let me know your feedback and any suggestions to change my suggestion - I'll be happy to build this functionality into 0.2, since I believe that this could really help make the slim-redux experience even smoother! So thanks a ton for bringing this up :)

**Comment by sagswe** [comment]
@aGuyNamedJonas , Firstly Thank you for the reply.   I liked the slimReduxReact(u r probably renaming it to connect soon) function with the object parameter. I can see the container component input state and output change triggers in one place. i am more worried towards better folder structure. I created the store in app folder which has bootstrapping part. It will be easy for me to see what goes inside the initial state in one place(rather than going through each reducer file in each feature folder and summing it). since we are specifying the subscription in slimReduxReact function, it makes sense to get that subscription slice as argument in reducer. and when the reducer is executed the  return value should replace only that piece of subscribed value, not the entire state.  As i said, I am worried about folder structure. I will have app folder for bootstrapping component and store creation js file. I will use the store in `<Provider>` . so store object wont be needed outside of  app folder. And containers folder will have containers and we create all inputs and outputs to the container component in 1 or 2 files max. 

I am not sure about the changeTrigger function you wrote above, does it require store object?. I prefer store object is needed only where we specify `<Provider>` componet. store object should not be needed in files where we specify the reducers.
And, any tentative date the slim-redux-react 0.2 version released?.

**Comment by aGuyNamedJonas** [comment]
> (u r probably renaming it to connect soon). 

Yessir! I think it's easier to remember that way - turning visual- into container-components like that is just really familiar from using `react-redux`.

> It will be easy for me to see what goes inside the initial state in one place(rather than going through each reducer file in each feature folder and summing it).

Sounds like a benefit slim-redux gives you which I never thought about, nice :)

> since we are specifying the subscription in slimReduxReact function, it makes sense to get that subscription slice as argument in reducer. and when the reducer is executed the return value should replace only that piece of subscribed value, not the entire state.  

Alright then - I'll build that mechanism into the change triggers! I think this will improve the experience of slim-redux, thanks again for bringing this idea up!

> I am not sure about the changeTrigger function you wrote above, does it require store object?. I prefer store object is needed only where we specify <Provider> componet. store object should not be needed in files where we specify the reducers.  

So the idea is that change triggers will not be objects anymore that you have to explicitly register in the store, rather you create them using `changeTrigger()`. The store object is not needed for `changeTrigger()` (so you can easily export them from wherever), the store object is only needed when you invoke the change trigger function that `changeTrigger()` returns.  
And in the case of slim-redux-react, it will use whatever store object is provided using the `<Provider>` component.   

So the store object is not needed where you specify the reducers. The big advantage you get from `changeTrigger()` is that it runs extensive validation functions (which you can see [here](https://github.com/aGuyNamedJonas/slim-redux/blob/0.2-release/src/changeTrigger.js#L3)) to make sure that when you define a change trigger it has the right format, the right data etc. But you can still define them independently of your store, only when you use them you have to (implicitly) specify which store object to use them on.

> And, any tentative date the slim-redux-react 0.2 version released?.

Let me say June 25th the latest! I'm almost done with everything, just needs some cleaning up here and there and I need to produce some updated documentation etc. My gf is gone on a promo job on a festival over the weekend, so I'll have even more time to finish this than usual :)  

Also if you want, send me your email address through a PM and I'll let you know when I need people to review the release + blog post etc. before putting it live!

**Comment by aGuyNamedJonas** [comment]
Was implemented in version 0.2. Thanks again for the inspiration @sagswe ! 🥇
